### PR TITLE
Remove WithCPK and thread options down the hierarchy

### LINKED ISSUE #7830: [P0] Storage: Remove WithCustomerProvidedKey methods
And move them to the `ClientOptions`.

### PR DESCRIPTION
**Pull request by tg-msft**
Fixes #7830

### CODE REVIEW

File: sdk/storage/Azure.Storage.Blobs/tests/TestExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Azure.Core;
+using Azure.Storage.Blobs;
+using Azure.Storage.Blobs.Models;
+using Azure.Storage.Blobs.Specialized;
+
+namespace Azure.Storage
```
Comment (JoshLove-msft) [inline sdk/storage/Azure.Storage.Blobs/tests/TestExtensions.cs]: Shouldn't this be in Test namespace?

File: sdk/storage/Azure.Storage.Blobs/src/BlobServiceClient.cs [hunk 2]
```diff
@@ -212,7 +223,7 @@ public BlobServiceClient(Uri serviceUri, TokenCredential credential, BlobClientO
         /// every request.
         /// </param>
         internal BlobServiceClient(Uri serviceUri, HttpPipelinePolicy authentication, BlobClientOptions options)
-            : this(serviceUri, authentication, options.Build(authentication), new ClientDiagnostics(options))
+            : this(serviceUri, authentication, options.Build(authentication), new ClientDiagnostics(options), options?.CustomerProvidedKey)
```
Comment (JoshLove-msft) [inline sdk/storage/Azure.Storage.Blobs/src/BlobServiceClient.cs]: Can options be null here? If so, there would already be a bug in the options.Build call.

**Comment by pakrym** [comment]
Do we have coverage for all the ctors? Especially around options being nullable? I've broken it once. Otherwise change LGTM

File: sdk/storage/Azure.Storage.Blobs/tests/TestExtensions.cs (hunk 1 above)
Comment (tg-msft) [reply sdk/storage/Azure.Storage.Blobs/tests/TestExtensions.cs]: It doesn't matter much and it's an existing partial class.  I'm trying to touch as few files as possible with all the big changes in flight.

File: sdk/storage/Azure.Storage.Blobs/src/BlobServiceClient.cs (hunk 2 above)
Comment (tg-msft) [reply sdk/storage/Azure.Storage.Blobs/src/BlobServiceClient.cs]: No - everyone who calls this does `options ?? new BlobClientOptions()`.

Comment (JoshLove-msft) [reply sdk/storage/Azure.Storage.Blobs/src/BlobServiceClient.cs]: So can you update to be options.CustomerProvidedKey?

**Review by amnguye — APPROVED** [review]

File: sdk/storage/Azure.Storage.Blobs/src/BlobServiceClient.cs (hunk 2 above)
Comment (tg-msft) [reply sdk/storage/Azure.Storage.Blobs/src/BlobServiceClient.cs]: I could, but this is still safer so I'd prefer to leave it.

**Comment by tg-msft** [comment]
Yes, we've got .ctor coverage here.  It might not be completely exhaustive, but it's pretty decent because we flow the client hierarchy a lot and call these particular overloads.

**Pull request merged**